src/theory/quantifiers/term_tuple_enumerator.cpp

```diff
@@ -0,0 +1,499 @@
+/*********************                                                        */
+/*! \file  term_tuple_enumerator.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Mikolas Janota
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of an enumeration of tuples of terms for the purpose
+ *of quantifier instantiation.
+ **/
+#include "theory/quantifiers/term_tuple_enumerator.h"
+
+#include <algorithm>
+#include <functional>
+#include <iterator>
+#include <map>
+#include <vector>
+
+#include "base/map_util.h"
+#include "base/output.h"
+#include "smt/smt_statistics_registry.h"
+#include "theory/quantifiers/index_trie.h"
+#include "theory/quantifiers/quant_module.h"
+#include "theory/quantifiers/term_util.h"
+#include "util/statistics_registry.h"
+
+namespace CVC4 {
+
+// TODO: kosher?
+template <typename T>
+static CVC4ostream& operator<<(CVC4ostream& out, const std::vector<T>& v)
+{
+  out << "[ ";
+  std::copy(v.begin(), v.end(), std::ostream_iterator<T>(out, " "));
+  return out << "]";
+}
+
+/** Tracing purposes, printing and mask that are of indices. */
+static void traceMaskedVector(const char* trace,
+                              const char* name,
+                              const std::vector<bool>& mask,
+                              const std::vector<size_t>& values)
+{
+  Assert(mask.size() == values.size());
+  Trace(trace) << name << " [ ";
+  for (size_t variableIx = 0; variableIx < mask.size(); variableIx++)
+  {
+    if (mask[variableIx])
+    {
+      Trace(trace) << values[variableIx] << " ";
+    }
+    else
+    {
+      Trace(trace) << "_ ";
+    }
+  }
+  Trace(trace) << "]" << std::endl;
+}
+
+namespace theory {
+namespace quantifiers {
+/**
```

Comment:
I generally prefer larger documentation to be in the h file, perhaps this could this go above the interface class?